packages/x-data-grid/src/hooks/features/pagination/useGridPaginationModel.ts

```diff
@@ -0,0 +1,272 @@
+  const pageCount = getPageCount(rowCount, pageSize);
+
+  if (
+    paginationModelProp &&
+    (paginationModelProp?.page !== paginationModel.page ||
+      paginationModelProp?.pageSize !== paginationModel.pageSize)
+  ) {
+    paginationModel = paginationModelProp;
+  }
+
+  const validPage = getValidPage(paginationModel.page, pageCount);
+  if (validPage !== paginationModel.page) {
+    paginationModel = { ...paginationModel, page: validPage };
+  }
+
+  throwIfPageSizeExceedsTheLimit(paginationModel.pageSize, signature);
+
+  return paginationModel;
+};
+
+/**
+ * @requires useGridFilter (state)
+ * @requires useGridDimensions (event) - can be after
+ */
+export const useGridPaginationModel = (
+  apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
+  props: Pick<
+    DataGridProcessedProps,
+    | 'paginationModel'
+    | 'onPaginationModelChange'
+    | 'autoPageSize'
+    | 'initialState'
+    | 'paginationMode'
+    | 'signature'
+    | 'rowHeight'
+  >,
+) => {
+  const logger = useGridLogger(apiRef, 'useGridPaginationModel');
+
+  const densityFactor = useGridSelector(apiRef, gridDensityFactorSelector);
+  const rowHeight = Math.floor(props.rowHeight * densityFactor);
+  apiRef.current.registerControlState({
+    stateId: 'paginationModel',
+    propModel: props.paginationModel,
+    propOnChange: props.onPaginationModelChange,
+    stateSelector: gridPaginationModelSelector,
+    changeEvent: 'paginationModelChange',
+  });
+
+  /**
+   * API METHODS
+   */
+  const setPage = React.useCallback<GridPaginationModelApi['setPage']>(
+    (page) => {
+      const currentModel = gridPaginationModelSelector(apiRef);
+      if (page === currentModel.page) {
+        return;
+      }
+      logger.debug(`Setting page to ${page}`);
+      apiRef.current.setPaginationModel({ page, pageSize: currentModel.pageSize });
+    },
+    [apiRef, logger],
+  );
+
+  const setPageSize = React.useCallback<GridPaginationModelApi['setPageSize']>(
+    (pageSize) => {
+      const currentModel = gridPaginationModelSelector(apiRef);
+      if (pageSize === currentModel.pageSize) {
+        return;
+      }
+
+      logger.debug(`Setting page size to ${pageSize}`);
+      apiRef.current.setPaginationModel({ pageSize, page: currentModel.page });
+    },
+    [apiRef, logger],
+  );
+
+  const setPaginationModel = React.useCallback<GridPaginationModelApi['setPaginationModel']>(
+    (paginationModel) => {
+      const currentModel = gridPaginationModelSelector(apiRef);
+      if (paginationModel === currentModel) {
+        return;
+      }
+      logger.debug("Setting 'paginationModel' to", paginationModel);
+
+      apiRef.current.setState((state) => ({
+        ...state,
+        pagination: {
+          ...state.pagination,
+          paginationModel: getDerivedPaginationModel(
+            state.pagination,
+            props.signature,
+            paginationModel,
+          ),
+        },
+      }));
+      apiRef.current.forceUpdate();
```

Comment:
Why add this one here? Are these still necessary?

Reply:
Not really, didn't remove it from the previous impl., removed it now!